python/src/iceberg/expression/literals.py

```diff
@@ -0,0 +1,502 @@
+# under the License.
+
+import datetime
+import struct
+import sys
+import uuid
+from abc import ABC, abstractmethod
+from decimal import ROUND_HALF_UP, Decimal
+from functools import singledispatch
+from typing import Generic, Optional, TypeVar, Union
+
+if sys.version_info >= (3, 8):
+    from functools import singledispatchmethod  # pragma: no cover
+else:
+    from singledispatch import singledispatchmethod  # pragma: no cover
+
+import pytz
+
+from iceberg.types import (
+    BinaryType,
+    BooleanType,
+    DateType,
+    DecimalType,
+    DoubleType,
+    FixedType,
+    FloatType,
+    IntegerType,
+    LongType,
+    Singleton,
+    StringType,
+    TimestampType,
+    TimestamptzType,
+    TimeType,
+    UUIDType,
+)
+
+EPOCH = datetime.datetime.utcfromtimestamp(0)
+T = TypeVar("T")
+
+
+class Literal(Generic[T], ABC):
+    """Literal which has a value and can be converted between types"""
+
+    def __init__(self, value: T):
+        if value is None:
+            raise TypeError(f"Invalid literal value: {value}")
+        self._value = value
+
+    @property
+    def value(self) -> T:
+        return self._value
+
+    @abstractmethod
+    def to(self, type_var):
+        ...  # pragma: no cover
+
+    def __repr__(self):
+        return f"{type(self).__name__}({self.value})"
+
+    def __str__(self):
+        return str(self.value)
+
+    def __eq__(self, other):
+        return self.value == other.value
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __lt__(self, other):
+        return self.value < other.value
+
+    def __gt__(self, other):
+        return self.value > other.value
+
+    def __le__(self, other):
+        return self.value <= other.value
+
+    def __ge__(self, other):
+        return self.value >= other.value
+
+
+@singledispatch
+def literal(value) -> Literal:
+    """
+    A generic Literal factory to construct an iceberg Literal based on python primitive data type
+    using dynamic overloading
+
+    Args:
+        value(python primitive type): the value to be associated with literal
+
+    Example:
+        from iceberg.expressions.literals import literal
+        >>> literal(123)
+        IntegerLiteral(123)
+    """
+    raise TypeError(f"Unimplemented Type Literal for value: {str(value)}")
+
+
+@literal.register(bool)
+def _(value: bool) -> Literal[bool]:
+    return BooleanLiteral(value)
+
+
+@literal.register(int)
+def _(value: int) -> Literal[int]:
+    """
+    Upgrade to long if python int is outside the JAVA_MIN_INT and JAVA_MAX_INT
+    """
+    if value < IntegerType.min or value > IntegerType.max:
+        return LongLiteral(value)
+    return IntegerLiteral(value)
+
+
+@literal.register(float)
+def _(value: float) -> Literal[float]:
+    """
+    Upgrade to double if python float is outside the JAVA_MIN_FLOAT and JAVA_MAX_FLOAT
+    """
+    if value < FloatType.min or value > FloatType.max:
+        return DoubleLiteral(value)
+    return FloatLiteral(value)
```

Comment:
This could just return `DoubleLiteral(value)`. If the type needed is a float, then expression binding will convert.